hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -97,126 +280,355 @@ impl<'t> YkPTBlockIterator<'t> {
+                    self.cur_loc = ObjLoc::MainObj(off);
+                    return Ok(block);
+                }
+            }
+
+            if !seg.vaddrs.contains(&vaddr) {
+                // The next instruction is outside of the current segment. Switch segment and make
+                // a new decoder for it.
+                seg = CODE_SEGS.seg(vaddr);
+                let seg_start_u64 = u64::try_from(seg.vaddrs.start).unwrap();
+                dis = iced_x86::Decoder::with_ip(64, seg.slice, seg_start_u64, 0);
+                dis.set_ip(u64::try_from(vaddr).unwrap());
+                reposition = true;
+            }
+
+            if reposition {
+                dis.set_position(vaddr - seg.vaddrs.start)
+                    .map_err(|_| HWTracerError::DisasmFail("failed to reposition".to_owned()))?;
+                reposition = false;
+            }
+
+            let inst = dis.decode();
+            match inst.flow_control() {
+                iced_x86::FlowControl::Next => (),
+                iced_x86::FlowControl::Return => {
+                    // We don't expect to see any 16-bit far returns.
+                    debug_assert!(is_ret_near(&inst));
+
+                    let ret_vaddr = if self.is_return_compressed()? {
+                        // This unwrap cannot fail if the CPU correctly implements compressed
+                        // returns.
+                        match self.comprets.pop().unwrap() {
+                            CompRetAddr::VAddr(vaddr) => vaddr,
+                            CompRetAddr::AfterCall(off) => {
+                                self.off_to_vaddr(&PHDR_MAIN_OBJ, off + 1)?
+                            }
+                        }
+                    } else {
+                        match self.cur_loc {
+                            ObjLoc::MainObj(off) => self.off_to_vaddr(&PHDR_MAIN_OBJ, off)?,
+                            ObjLoc::OtherObjOrUnknown(opt_vaddr) => match opt_vaddr {
+                                Some(vaddr) => vaddr,
+                                None => unreachable!(),
+                            },
+                        }
+                    };
+                    dis.set_ip(u64::try_from(ret_vaddr).unwrap());
+                    reposition = true;
+                }
+                iced_x86::FlowControl::IndirectBranch | iced_x86::FlowControl::IndirectCall => {
+                    self.seek_tip()?;
+                    let vaddr = match self.cur_loc {
+                        ObjLoc::MainObj(off) => self.off_to_vaddr(&PHDR_MAIN_OBJ, off)?,
+                        ObjLoc::OtherObjOrUnknown(opt_vaddr) => match opt_vaddr {
+                            Some(vaddr) => vaddr,
+                            None => unreachable!(),
+                        },
+                    };
+
+                    if inst.flow_control() == iced_x86::FlowControl::IndirectCall {
+                        if inst.next_ip() as usize == vaddr {
```

Comment:
Is the `as` always safe? If so, we should briefly say why. [I dislike `as` in Rust!]

Reply:
It is safe for x86_64 (it's `u64` to `usize`), but this is a leftover from hacking. In principle i aim to never use `as`, but during hacking I find it too tedious to do the `usize::try_from(...).unwrap()` dance. Will fix later.

Reply:
Fixed in 98d683e662382b103e3ad05e1d5dcacd5cacfdc9